### PR TITLE
#49 object support

### PR DESCRIPTION
**Pull request by burntcookie90**
Initial implementation of #49

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -190,6 +195,11 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
         firstMember = false
       }
 
+      // Companion object
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Delete, not useful

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 2]
```diff
@@ -101,6 +102,10 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
         codeWriter.emitModifiers(modifiers, setOf(PUBLIC))
         if (kind == Kind.ANNOTATION) {
           codeWriter.emit("annotation class %L", name)
+        } else if (kind == Kind.OBJECT) {
+          codeWriter.emit("object class %L", name)
+        } else if (kind == Kind.COMPANION) {
+          codeWriter.emit("companion object")
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: These can have custom names. Do we care?

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 3]
```diff
@@ -311,6 +330,13 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
       return this
     }
 
+    fun companionObject(companionObject: TypeSpec): Builder {
+      check(kind == Kind.CLASS || kind == Kind.INTERFACE) { "$kind can't have a companion object" }
+      require(companionObject.kind == Kind.COMPANION) { "expect a companion object class but was $kind "}
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Expected

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 4]
```diff
@@ -323,7 +349,9 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     }
 
     fun superclass(superclass: TypeName): Builder {
-      check(kind == Kind.CLASS) { "only classes have super classes, not $kind" }
+      check(kind == Kind.CLASS || kind == Kind.OBJECT || kind == Kind.COMPANION) {
+        "only classes have super classes, not $kind"
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Can have

File: src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 5]
```diff
@@ -2143,6 +2143,106 @@ class TypeSpecTest {
...
+        .addInitializerBlock(CodeBlock.builder().build())
+        .addFun(FunSpec.builder("test")
+            .addModifiers(KModifier.PUBLIC)
+            .build())
+        .build()
+
+    assertThat(toString(type)).isEqualTo("""
+        |package com.squareup.tacos
+        |
+        |object class MyObject {
```
Comment (JakeWharton) [inline src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: Shouldn't emit `class` here

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 2 above)
Comment (burntcookie90) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: good question, it's not a feature i've ever used, but it wouldn't be hard to include

Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Skip for now. Would be a good followup

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 6]
```diff
@@ -323,7 +348,9 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     }
 
     fun superclass(superclass: TypeName): Builder {
-      check(kind == Kind.CLASS) { "only classes have super classes, not $kind" }
+      check(kind == Kind.CLASS || kind == Kind.OBJECT || kind == Kind.COMPANION) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: nice

**Pull request merged**